### DISCUSSION TITLE
What is the current behaviour of UDP receiver, when dealing with out-of-order packets? (5.11.x / 5.12.x release)

### DISCUSSION THREAD

**Discussion by chengguizi** [eclipse-ecal/ecal#1283] [Q&A] [answered]
I just notice there is some new development on UDP design with eCAL, pretty interesting.
https://github.com/eclipse-ecal/ecal/pull/1271

It reminds me a question releated to the design of the UDP pub/sub of eCAL. May I know if the current release version of eCAL handles out-of-order UDP packages, during the reception of the eCAL message (say ~1MB big)? In other words, if packages are received out of order, will the design tolerate that and still recover the right full eCAL message?

Now I see there is a de-fragmentation talked about in this PR. Does it mean it is improving the situration of the out-of-order UDP package reception?

Thanks!

**Comment by FlorianReimold** [comment]
Hi @chengguizi,

- The current PRs re-organize the current UDP code, but don't introduce new features. The goal is to reduce the coupling of eCAL to its own UDP implementation. We are currently working on a new implementation of the UDP code that will have a proper API and can also be used standalone. The new implementation will be easier to test and maintain.
- 1 MiB packages will always be critical, as an ethernet frame only fits 1500 bytes. UDP is a fast but unrealiable protocol, meaning that if one of those 1500 byte fragments is lost, your entire 1 MiB message is gone. I don't say that it cannot work, but I would recommend using the eCAL TCP layer for those large messages.
- Out of order fragments are handled both in the current and in the future UDP implementation. Out-of-order fragements will be buffered and reassembled once the entire message is complete
- Can you specify what we need to improve with out-of-order "packages"? Does your term "package" refer to a single fragment (i.e. a UDP datagram) or the reassembled eCAL message?

Kind regards
Florian

**Reply by chengguizi** [reply]
Hi @FlorianReimold thanks for your detailed answer. Yes, I am fully aware of the unreliable nature of the UDP. In our test, on LAN network, loss of UDP datagram seems to be quite low probability, but out of order of UDP datagram does happen. Hence my question earlier on.
Glad to know that the out of order UDP datagram is handled. Specifically, would this case be properly handled:
the content UDP datagram arrives before the header UDP datagram for a give eCAL full message.
https://github.com/eclipse-ecal/ecal/blob/0f16b39e964ebe5e2c64fbb2a8ddeea519f19f6b/ecal/core/src/io/udp/ecal_udp_sample_receiver.cpp#L316-L317

from there, it seems to suggest that, if there are any content UDP datagram arrives before the header UDP datagram, the whole message is lost.

Am I deducing this correctly? Seems like something to handle more nicely.

**Reply by rex-schilasky** [reply]
Hi @chengguizi, yes your deduction is fully correct. The current implementation will not resort out of order packages in any kind. So if a data package arrives before the initial message header package or if data packages arrive in the wrong order, the whole message is lost. My understanding of `package` is the higher level eCAL udp fragment, not the UDP datagram.

**Reply by chengguizi** [reply]
Hi @rex-schilasky May I clarify further what is the difference between the invidisual eCAL udp fragment and UDP datagram. From my limited inspection, the eCAL udp fragment is essentially one UDP packet / datagram. Or more precisely, the eCAL udp fragment is inside the data section of one udp packet / datagram. Am I correct?

Meaning to say, the following line is to receive the content of UDP package, and cast it later to a eCAL udp fragment

https://github.com/eclipse-ecal/ecal/blob/0f16b39e964ebe5e2c64fbb2a8ddeea519f19f6b/ecal/core/src/io/udp/ecal_udp_sample_receiver.cpp#L140

Am I still correct saying all these?

**Reply by chengguizi** [reply]
If so, could we implement some simple logic here to buffer the "advanced" eCAL udp fragment that arrive before the header, and then release them after the header arrives. This way we should have a much higher chance of recovering big UDP data, as long as there is no actual packet loss (which is quite possible if we work on LAN i think)

**Reply by FlorianReimold** [reply, marked as answer]
For fragmented messages, the current eCAL v5 UDP protocol requires an entire UDP datagram to be transmitted for telling the receiver about the fragmented package. This datagram (called "header" in the current implementation) cannot contain any actual user payload. The payload is transmitted with the next datagrams (called "payload" in the current implementation). And yes, it looks like an out-of-order header would cause the message to be lost. This can be considered to be a bug.

The new implementation that I am working on will:
- obviously handle out of order datagrams by simply storing them in a map. And yes, it does that for those "header" and the "payload" datagrams
- properly look at the sender endpoint when reassembling fragmented messages
- require less threads and - more importantly - no polling of sockets
- Probably introduce a new protocol that doesn't require a "header" to be sent alone for fragmented messages (because that's inefficient)

**Reply by chengguizi** [reply]
This would be a fantastic solution. Very looking forward to it!

**Reply by chengguizi** [reply]
@FlorianReimold Wondering if this "bug" fix is being worked on towards 5.13? Or would be available in the further future?

**Reply by FlorianReimold** [reply]
Good and bad news:
- Yes, it is being worked on
- No, I don't think it will make it in 5.13 :-(

The reason is that converting the special Windows NPCAP based UDP receiver to the new structure will take some additional time...

**Reply by chengguizi** [reply]
No worries! Will continue to look forward then :)
